### PR TITLE
[ParticleMechanicsApplication] Adding implicit mod-camclay CL

### PR DESCRIPTION
**Pull request by bodhinandach**
@antonialarese @iiaconeta 
This PR is to add a new constitutive law, the Modified Cam Clay, to the ParticleMechanicsApplication. I have done some simple tests to check the performance of the codes. However, no concrete validations are conducted yet. I just don't want the PR becomes too large to review and will do the follow-up tests later.

This law considers associative return mapping for large deformation and I appended it to the hencky_plastic_3d_law that is used also by the Mohr-Coulomb family law.

### CODE REVIEW

**Review by loumalouomega — DISMISSED** [review]
Remove Tetgen plis

**Comment by bodhinandach** [comment]
@loumalouomega I am not sure why that tetgen is always coming into my commit... Any idea how to remove that? I will message you.

**Comment by bodhinandach** [comment]
@loumalouomega removed.

File: applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/MPM_flow_rule.hpp [hunk 1]
```diff
@@ -226,6 +230,8 @@ class KRATOS_API(PARTICLE_MECHANICS_APPLICATION) MPMFlowRule
             std::cout<<" Internal Variables "<<std::endl;
             std::cout<<" EquivalentPlasticStrain: "<<EquivalentPlasticStrain<<std::endl;
             std::cout<<" DeltaPlasticStrain: "<<DeltaPlasticStrain<<std::endl;
+            std::cout<<" AccumulatedPlasticVolumetricStrain: "<<AccumulatedPlasticVolumetricStrain<<std::endl;
```
Comment (loumalouomega) [inline applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/MPM_flow_rule.hpp:233]: KRATOS_INFO?

File: applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/borja_cam_clay_plastic_flow_rule.cpp [hunk 2]
```diff
@@ -0,0 +1,777 @@
...
+    Vector& rPrincipalStrain, unsigned int& region, Vector& rPrincipalStressUpdated)
+{
+
+    // Calculate stress return in principal stress
+    // The flow rule is written for associated plasticity assumption
+    // Refer to paper by (Borja, 1998) for the theoretical description
+    bool converged = false;
+
+    // Initiate Main Matrices and Vectors for Newton Iteration
+    Vector UnknownVector, DeltaUnknownVector;
```
Comment (loumalouomega) [inline applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/borja_cam_clay_plastic_flow_rule.cpp:196]: If the size is fix I recommend to use bounded matrices and vectors

File: applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/borja_cam_clay_plastic_flow_rule.cpp [hunk 3]
```diff
@@ -0,0 +1,777 @@
...
+        K_p *= 1.0 / (OtherSlope-SwellingSlope);
+
+        // Calculate RHS Vector
+        RHSVector(0) = UnknownVector(0) - TrialVolumetricStrain + UnknownVector(3) * mStateFunctionFirstDerivative(0);
+        RHSVector(1) = UnknownVector(1) - TrialDeviatoricStrain + UnknownVector(3) * mStateFunctionFirstDerivative(1);;
+        RHSVector(2) = mStateFunction;
+
+        // Calculate RHS Norm (Residual Norm)
+        if (counter == 0) InitialNormResidual = norm_2(RHSVector);
+        double CurrentNormResidual = norm_2(RHSVector);
```
Comment (loumalouomega) [inline applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/borja_cam_clay_plastic_flow_rule.cpp:253]: variables declared inside a region->snake_case, input variables and member variables->CammelCase

File: applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/borja_cam_clay_plastic_flow_rule.cpp [hunk 4]
```diff
@@ -0,0 +1,777 @@
...
+    Matrix Tensor_1xN = ZeroMatrix(6,6);
+    for (unsigned int i = 0; i<3; ++i) 
+    {
+        for (unsigned int j = 0; j<3; ++j) 
+        {
+            Tensor_1xN(i,j) = DirectionVector(j);
+        }
+    }
+
+    Matrix Tensor_Nx1 = ZeroMatrix(6,6);
```
Comment (loumalouomega) [inline applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/borja_cam_clay_plastic_flow_rule.cpp:603]: Idem, if the size is fix use bounded matrices

File: applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/borja_cam_clay_plastic_flow_rule.cpp [hunk 5]
```diff
@@ -0,0 +1,777 @@
...
+    {
+        for (unsigned int j = 0; j<3; ++j) 
+        {
+            Tensor_Nx1(i,j) = DirectionVector(i);
+        }
+    } 
+
+    // Perform check in case DeviatoricStrain == 0
+    double DeviatoricQ_by_DeviatoricStrain; 
+    if (fabs(DeviatoricStrain) < 1.e-9 ) {DeviatoricQ_by_DeviatoricStrain = DeviatoricQ / 1.e-9;}
```
Comment (loumalouomega) [inline applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/borja_cam_clay_plastic_flow_rule.cpp:614]: std::abs

File: applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/borja_cam_clay_plastic_flow_rule.cpp [hunk 6]
```diff
@@ -0,0 +1,777 @@
...
+    else{DeviatoricQ_by_DeviatoricStrain = DeviatoricQ / DeviatoricStrain;}  
+
+    // Compute Consistent Tangent Stiffness matrix in principal space
+    Matrix DepcP = ZeroMatrix(6,6);
```
Comment (loumalouomega) [inline applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/borja_cam_clay_plastic_flow_rule.cpp:618]: bounded

File: applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/borja_cam_clay_plastic_flow_rule.cpp [hunk 7]
```diff
@@ -0,0 +1,777 @@
...
+    K_p *= 1.0 / (OtherSlope-SwellingSlope);
+    mMaterialParameters.PlasticHardeningModulus = K_p;
+
+    // Update Consistency Parameter
+    mMaterialParameters.ConsistencyParameter = rConsistencyParameter;
+}
+
+double BorjaCamClayPlasticFlowRule::GetSmoothingLodeAngle()
+{
+    return 29.0*GetPI()/180.0;
```
Comment (loumalouomega) [inline applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/borja_cam_clay_plastic_flow_rule.cpp:752]: @AlejandroCornejo look another magic 29

Comment (loumalouomega) [reply applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/borja_cam_clay_plastic_flow_rule.cpp:752]: @bodhinandach do you know why 29?

File: applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/borja_cam_clay_plastic_flow_rule.hpp [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+    {
+        double PreconsolidationPressure;
+        double PlasticHardeningModulus;
+        double ConsistencyParameter;
+
+    public:
+        void PrintInfo()
+        {
+            std::cout << "PreconsolidationPressure   = " <<  PreconsolidationPressure  << std::endl;
+            std::cout << "PlasticHardeningModulus   = " <<  PlasticHardeningModulus  << std::endl;
```
Comment (loumalouomega) [inline applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/borja_cam_clay_plastic_flow_rule.hpp:89]: KRATOS_INFO

File: applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/mc_plastic_flow_rule.hpp [hunk 9]
```diff
@@ -76,19 +76,16 @@ class MCPlasticFlowRule
...
         double FrictionAngle;
         double DilatancyAngle;
 
     public:
         void PrintInfo()
         {
-            std::cout << "YoungModulus   = " << YoungModulus   << std::endl;
-            std::cout << "PoissonRatio   = " << PoissonRatio   << std::endl;
             std::cout << "Cohesion       = " << Cohesion       << std::endl;
             std::cout << "FrictionAngle  = " << FrictionAngle  << std::endl;
```
Comment (loumalouomega) [inline applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/mc_plastic_flow_rule.hpp:90]: KRATOS_INFO

File: applications/ParticleMechanicsApplication/custom_constitutive/hardening_laws/exponential_strain_softening_law.cpp [hunk 10]
```diff
@@ -71,31 +71,27 @@ double& ExponentialStrainSofteningLaw::CalculateHardening(double &rHardening, co
...
-            rHardening = -1.0 * Beta * (PeakFrictionAngle - ResidualFrictionAngle) * std::exp( -1.0 * Beta * rAlpha );
-        }
-
-        else if (rThisVariable == INTERNAL_DILATANCY_ANGLE)
-        {
-            const double PeakDilatancyAngle = GetProperties()[INTERNAL_DILATANCY_ANGLE];
-            const double ResidualDilatancyAngle = GetProperties()[INTERNAL_DILATANCY_ANGLE_RESIDUAL];
-            rHardening = -1.0 * Beta * (PeakDilatancyAngle - ResidualDilatancyAngle) * std::exp( -1.0 * Beta * rAlpha );
-        }
+        const double PeakFrictionAngle = GetProperties()[INTERNAL_FRICTION_ANGLE];
```
Comment (loumalouomega) [inline applications/ParticleMechanicsApplication/custom_constitutive/hardening_laws/exponential_strain_softening_law.cpp:85]: snake_case https://github.com/KratosMultiphysics/Kratos/wiki/Style-Guide

File: applications/ParticleMechanicsApplication/custom_constitutive/hencky_borja_cam_clay_3D_law.cpp [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+namespace Kratos
+{
+
+//******************************CONSTRUCTOR*******************************************
+//************************************************************************************
+
+HenckyBorjaCamClayPlastic3DLaw::HenckyBorjaCamClayPlastic3DLaw()
+    : HenckyElasticPlastic3DLaw()
+{
+    mpHardeningLaw   = HardeningLaw::Pointer( new CamClayHardeningLaw() );
```
Comment (loumalouomega) [inline applications/ParticleMechanicsApplication/custom_constitutive/hencky_borja_cam_clay_3D_law.cpp:32]: Use make_shared instead, is more efficient than new

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Please take into consideration my suggestions

File: applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/borja_cam_clay_plastic_flow_rule.cpp (hunk 7 above)
Comment (bodhinandach) [reply applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/borja_cam_clay_plastic_flow_rule.cpp:752]: Hahaha I dont use LodeAngle here. But I guess it's empirical? :laughing:

Comment (AlejandroCornejo) [reply applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/borja_cam_clay_plastic_flow_rule.cpp:752]: haha i'll ask S.Oller eventually about that..

Comment (loumalouomega) [reply applications/ParticleMechanicsApplication/custom_constitutive/flow_rules/borja_cam_clay_plastic_flow_rule.cpp:752]: It's magic

**Review by loumalouomega — APPROVED** [review]

**Pull request merged**